File: course-schedule-ii/course-schedule-ii.java
class Solution {
    Map<Integer, List<Integer>> vertices;
    Boolean hasCycle = false;
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        vertices = new HashMap<Integer, List<Integer>>();
        for(var edge: prerequisites) {
            var list = vertices.getOrDefault(edge[0], new ArrayList<Integer>());
            list.add(edge[1]);
            vertices.put(edge[0], list);
        }
        var visited = new HashSet<Integer>();
        var stack = new ArrayDeque<Integer>();
        var visitedCurr = new HashSet<Integer>();
        for(int i=0; i<numCourses; i++) {
            if(!visited.contains(i)) 
                topSort(i, visited, stack, visitedCurr);
        }
        if (hasCycle) {
            print("hasCycle");
            return new int[0];
        }
        else {
            int[] res1 = new int[stack.size()];
            int count = stack.size() - 1;
            for(var a: stack) {
                res1[count] = a;
                count--;
            }
            return res1;
        }
    }
    
    public void topSort(int vertex, Set<Integer> visited, Deque<Integer> stack, Set<Integer> visitedCurr) {
        visited.add(vertex);
        visitedCurr.add(vertex);
        for(var nbr: vertices.getOrDefault(vertex, new ArrayList<Integer>())) {
            if(!visited.contains(nbr)) {
                topSort(nbr, visited, stack, visitedCurr);
            }
            else if(visitedCurr.contains(nbr)) {
                hasCycle = true;   
            }
        }
        visitedCurr.remove(vertex);
        stack.addFirst(vertex);
    }
    
    public static void print(Object o, String sep) {
        System.out.print(o + sep);
    }
    public static void print(Object o) {
        print(o, "\n");
    }
}

File: course-schedule-ii/course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        def rec(key, curr_visited):
            nonlocal visited, topoOrder, hasCycle, graph
            if hasCycle: return
            visited.add(key)
            curr_visited.add(key)
            for nbr in graph[key]:
                if nbr not in visited:
                    rec(nbr, curr_visited)
                elif nbr in curr_visited:
                    hasCycle = True
            if not hasCycle:
                topoOrder.append(key)
            curr_visited.remove(key)
            
                    
            
        
        graph = defaultdict(list)
        for prq in prerequisites: graph[prq[0]] += [prq[1]]
        
        
        topoOrder, hasCycle = [], False
        
        visited, curr_visited = set(), set()
        
        for i in range(numCourses):
            if i not in visited:
                rec(i, curr_visited)
            
        if not hasCycle:
            return topoOrder
        else:
            return []